components/channel/test/channel_test.c

```diff
@@ -0,0 +1,398 @@
+  TEST_ASSERT_LIST_EMPTY (&ch1.unique);
+  TEST_ASSERT_LIST_EMPTY (&ch2.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch0.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch1.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+
+  channel_register(&ch3);
+  TEST_ASSERT_LIST_LENGTH(&ch0.unique, 3);
+  TEST_ASSERT_LIST_EMPTY (&ch1.unique);
+  TEST_ASSERT_LIST_EMPTY (&ch2.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch3.unique, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch0.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch1.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+  TEST_ASSERT_LIST_EMPTY (&ch3.same);
+
+  channel_register(&ch4);
+  TEST_ASSERT_LIST_LENGTH(&ch0.unique, 3);
+  TEST_ASSERT_LIST_EMPTY (&ch1.unique);
+  TEST_ASSERT_LIST_EMPTY (&ch2.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch3.unique, 3);
+  TEST_ASSERT_LIST_EMPTY (&ch4.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch0.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch1.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch3.same, 2);
+  TEST_ASSERT_LIST_LENGTH(&ch4.same, 2);
+
+  channel_register(&ch5);
+  TEST_ASSERT_LIST_LENGTH(&ch0.unique, 4);
+  TEST_ASSERT_LIST_EMPTY (&ch1.unique);
+  TEST_ASSERT_LIST_EMPTY (&ch2.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch3.unique, 4);
+  TEST_ASSERT_LIST_EMPTY (&ch4.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch5.unique, 4);
+  TEST_ASSERT_LIST_LENGTH(&ch0.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch1.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch3.same, 2);
+  TEST_ASSERT_LIST_LENGTH(&ch4.same, 2);
+  TEST_ASSERT_LIST_EMPTY (&ch5.same);
+
+  TEST_ASSERT(channel_debug_resetRoot(&ch0));
+}
+
+TEST_CASE("channel_unregister", "[channel]")
+{
+  Channel ch0, ch1, ch2, ch3, ch4, ch5;
+  char identifier0[] = "test_channel0";
+  char identifier1[] = "test_channel1";
+  char identifier2[] = "test_channel2";
+  void *ctx = (void*)0x1;
+  BaseType_t flags = 0x2;
+  Channel_callback callback = &returnPdPASS;
+  channel_init(&ch0, identifier0, ctx, flags, callback);
+  channel_init(&ch1, identifier0, ctx, flags, callback);
+  channel_init(&ch2, identifier0, ctx, flags, callback);
+  channel_init(&ch3, identifier1, ctx, flags, callback);
+  channel_init(&ch4, identifier1, ctx, flags, callback);
+  channel_init(&ch5, identifier2, ctx, flags, callback);
+
+  channel_register(&ch0);
+  channel_register(&ch1);
+  channel_register(&ch2);
+  channel_register(&ch3);
+  channel_register(&ch4);
+  channel_register(&ch5);
+
+  TEST_ASSERT_LIST_LENGTH(&ch0.unique, 4);
+  TEST_ASSERT_LIST_EMPTY (&ch1.unique);
+  TEST_ASSERT_LIST_EMPTY (&ch2.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch3.unique, 4);
+  TEST_ASSERT_LIST_EMPTY (&ch4.unique);
+  TEST_ASSERT_LIST_LENGTH(&ch5.unique, 4);
+  TEST_ASSERT_LIST_LENGTH(&ch0.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch1.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch3.same, 2);
+  TEST_ASSERT_LIST_LENGTH(&ch4.same, 2);
+  TEST_ASSERT_LIST_EMPTY (&ch5.same);
+
+  channel_unregister(&ch3);
```

Comment:
This covers only the most complex path. Maybe add tests for unregister single unique and same channels